File: frontend/app/types/incident.ts
export interface IncidentResponse {
  id: number
  onCallPeriodId: number
  name: string
  startDateTime: string // ISO-8601 LocalDateTime e.g. "2025-06-03T02:30:00"
  endDateTime: string // ISO-8601 LocalDateTime e.g. "2025-06-03T04:15:00"
  createdAt: string // ISO-8601 Instant
}

export interface CreateIncidentRequest {
  onCallPeriodId: number
  name: string
  startDateTime: string // "2025-06-03T02:30:00"
  endDateTime: string // "2025-06-03T04:15:00"
}

export interface UpdateIncidentRequest {
  name: string
  startDateTime: string
  endDateTime: string
}

File: backend/src/main/java/com/github/marcelorodrigo/dutytracker/domain/Incident.java
package com.github.marcelorodrigo.dutytracker.domain;

import com.github.marcelorodrigo.dutytracker.domain.exceptions.InvalidIncidentException;
import java.time.LocalDateTime;

public record Incident(
        Long id,
        Long onCallPeriodId,
        String name,
        LocalDateTime startDateTime,
        LocalDateTime endDateTime,
        LocalDateTime createdAt) {

    public Incident {
        if (onCallPeriodId == null || onCallPeriodId <= 0) {
            throw new InvalidIncidentException("onCallPeriodId must be a positive number");
        }
        if (name == null || name.isBlank()) {
            throw new InvalidIncidentException("name is required");
        }
        if (startDateTime == null || endDateTime == null) {
            throw new InvalidIncidentException("Incident startDateTime and endDateTime are required");
        }
        if (endDateTime.isBefore(startDateTime.plusMinutes(1))) {
            throw new InvalidIncidentException("Incident endDateTime must be at least 1 minute after startDateTime");
        }
    }

    public static Incident create(
            Long onCallPeriodId,
            String name,
            LocalDateTime startDateTime,
            LocalDateTime endDateTime,
            LocalDateTime createdAt) {
        return new Incident(null, onCallPeriodId, name, startDateTime, endDateTime, createdAt);
    }

    public Incident withDetails(String name, LocalDateTime startDateTime, LocalDateTime endDateTime) {
        return new Incident(id, onCallPeriodId, name, startDateTime, endDateTime, createdAt);
    }
}
